src/api-service/__app__/onefuzzlib/azure/nsg.py

```diff
@@ -0,0 +1,309 @@
+#!/usr/bin/env python
+#
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+import logging
+import os
+from typing import Any, Dict, List, Optional, Union, cast
+
+from azure.core.exceptions import ResourceNotFoundError
+from azure.mgmt.network.models import (
+    NetworkInterface,
+    NetworkSecurityGroup,
+    SecurityRule,
+    SecurityRuleAccess,
+    SecurityRuleProtocol,
+)
+from msrestazure.azure_exceptions import CloudError
+from onefuzztypes.enums import ErrorCode
+from onefuzztypes.models import Error
+from onefuzztypes.primitives import Region
+from pydantic import BaseModel, validator
+
+from .creds import get_base_resource_group
+from .network_mgmt_client import get_network_client
+
+
+def get_nsg(name: str) -> Optional[NetworkSecurityGroup]:
+    resource_group = get_base_resource_group()
+
+    logging.debug("getting nsg: %s", name)
+    network_client = get_network_client()
+    try:
+        return cast(
+            NetworkSecurityGroup,
+            network_client.network_security_groups.get(resource_group, name),
+        )
+    except (ResourceNotFoundError, CloudError) as err:
+        logging.debug("nsg %s does not exist %s", name, err)
+        return None
+
+
+def create_nsg(name: str, location: Region) -> Union[None, Error]:
+    resource_group = get_base_resource_group()
+
+    logging.info("creating nsg %s:%s:%s", resource_group, location, name)
+    network_client = get_network_client()
+
+    params: Dict = {
+        "location": location,
+    }
+
+    if "ONEFUZZ_OWNER" in os.environ:
+        params["tags"] = {"OWNER": os.environ["ONEFUZZ_OWNER"]}
+    try:
+        network_client.network_security_groups.begin_create_or_update(
+            resource_group, name, params
+        )
+    except (ResourceNotFoundError, CloudError) as err:
+        if "The request failed due to conflict with a concurrent request" in str(err):
```

Comment:
How do we know that the concurrent request was (1) a request to create the same NSG, (2) one that succeeded? In other words, why are we justified in returning `None`?

Example concern: we're invoking `begin_create_or_update()`. Can this request fail due to a concurrent request to _delete_ the NSG? If that were true, the concurrently-serviced `delete` would delete the NSG, but we'd return `None` here, implying the NSG creation request was successful.

Reply:
https://github.com/microsoft/onefuzz/blob/19ecba43cb964e4c3f48b30a8b285d09efbc5d31/src/api-service/__app__/onefuzzlib/azure/vm.py#L114

I'm using vm.py as template. Since that code been validated and battle-tested, nsg is following same pattern

Reply:
Opened work item: 
https://github.com/microsoft/onefuzz/issues/1349

Reply:
https://github.com/microsoft/onefuzz/issues/1350